### DISCUSSION TITLE
Problems with unusual rate limits

### DISCUSSION THREAD

**Discussion by afanjul** [postalserver/postal#2226] [Help with using Postal]
Hi @willpower232 , we keep having problems with google rate limits, and we are concluding that the problem could be caused/worsen due to the the retries of the soft-fail emails, I mean: 

If for any reason we fall in a "unusual rate limit" for our domain in gmail, postal will keep sending more and more emails until reach the hard-fail (we have even lower the attempts to 9), but each try of sending a new email is worsen the "gmail rate limit" because you are sending more and more emails... 

I know that you are not thinking in make a throttling general system, but could you give me some clues or pseudocode just to increase the retry time of soft-fails in case of a "substring match" of the failure message? I mean, if the failure message contains the word "rate", set the retry time of the next email to 10x times the default one (for example).

I think that the code to modify is here: https://github.com/postalserver/postal/blob/0dc6824a8f0315ea42b08f7e6812b821b62489c9/lib/postal/smtp_sender.rb#L169 
or maybe here? https://github.com/postalserver/postal/blob/21a8d890459958375d4a49a5b7f31f4900a9e8b1/app/models/queued_message.rb#L96
Im not sure... 

Could you please help me just with a pseudocode or some code you could write here? (I cant test it and modify afterwards)

Thanks in advance

**Comment by willpower232** [comment]
I am not really aware of any rate limit from gmail, what is the exact error message you are seeing?

It is likely that the message contains a unique-ish error code which would be much easier to detect than the word "rate".

If you are ticking off one of their limits now then delaying that by 10x probably isn't going to do much other than trigger off their warning 10x in the future. You might want to consider how many messages you're sending and whether you can batch them up and spread them out.

It might also be worth checking mxtoolbox.com and mail-tester.com to see if you have any more general improvements you can make to your Postal installation.

**Comment by afanjul** [comment]
Hi, thanks for your quick response, effectively Gmail has a "rate limit" policy, probably depending on the average sending rate you usually has. Some times (we don't know why) that limit triggers off so the emails get "stuck" in postal with hard-fails/held. 

The error message is this:

```
421-4.7.28 [212.83.191.250      15] Our system has detected an unusual rate of
421-4.7.28 unsolicited mail originating from your IP address. To protect our
421-4.7.28 users from spam, mail sent from your IP address has been temporarily
421-4.7.28 rate limited. Please visit
421-4.7.28  https://support.google.com/mail/?p=UnsolicitedRateLimitError to
421 4.7.28 review our Bulk Email Senders Guidelines. a17-20020adfed11000000b00228611a7e24si4129251wro.136 - gsmtp
```

although there are other error codes similar to this but from yahoo, etc.

So finally we have modify the code so that if the "soft fail" is caused by a "rate limit" the retry time quite increased....

We have changed this function: https://github.com/postalserver/postal/blob/0dc6824a8f0315ea42b08f7e6812b821b62489c9/lib/postal/smtp_sender.rb#L169

Like this:

```
rescue Net::SMTPServerBusy, Net::SMTPAuthenticationError, Net::SMTPSyntaxError, Net::SMTPUnknownError, Net::ReadTimeout => e
    log "#{e.class}: #{e.message}"
    result.type = 'SoftFail'
    result.retry = true
    result.details = "Temporary SMTP delivery error when sending to #{destination_host_description}"
    result.output = e.message
    if e.to_s =~ /(\d+) seconds/
        result.retry = $1.to_i + 10
    elsif e.to_s =~ /(\d+) minutes/
        result.retry = ($1.to_i * 60) + 10
    elsif e.to_s =~ /(421-4\.7\.28|421\s4\.7\.0|451\s4\.7\.651|451\s4\.7\.652|451\s4\.7\.653)/
        result.retry = rand(5400..9000)
    end

    safe_rset
```

so if the smtp error message is related with the "rate limits" codes, the retry time is a number of seconds between 5400 and 9000. It seems to works fine... the only caveat is that the queue could be fill up of waiting messages a bit quicker...but I think is a minor problem....

how do you see it?

**Reply by willpower232** [reply]
that all makes sense I guess, might be worth running some of your messages through mail-tester.com in case there is something you have forgotten, or your content sounds spammy

If you have problems with the queue, you should be able to scale the workers to deal with it, time will tell.
